src/aspire/source/__init__.py

```diff
@@ -430,9 +425,9 @@ def __init__(self, im, metadata=None):
         :param metadata: A Dataframe of metadata information corresponding to this ImageSource's images
         """
         super().__init__(L=im.res, n=im.n_images, dtype=im.dtype, metadata=metadata, memory=None)
-        self._im = im
+        self._cached_im = im
 
     def _images(self, start=0, num=np.inf, indices=None):
         if indices is None:
             indices = np.arange(start, min(start + num, self.n))
-        return self._im[indices]
+        return self._cached_im[indices]
```

Comment:
This code is never run, since `ImageSource.images` wil just pull images from `self._cached_im`, correct? Might be good to just leave this empty with a comment to that effect.

Reply:
In simulation and RelionSource classes, `_images` is overrode to return the images from projections of 3D map or mrc files. I think this one is never called.

Reply:
@junchaoxia .. this is the `_images` method in `ArrayImageSource`. Yes..for `ArrayImageSource` it's sufficient to initialize `_cached_im` and not override this method at all (and leave a comment explaining why..). Coming up.

Reply:
@vineetbansal Thanks. I forgot there is one in the `ArrayImageSource` class which is designed for the non-mutable nature of `ImageSource` class.